### PR TITLE
Moving some types/APIs from S.IO and S.Reflection.Primitives to S.P.CoreLib, to support new Reflection refactoring

### PR DESCRIPTION
**Pull request by SedarG**
@weshaggard , I'm ready to merge my changes to dev/api. Could you please take a look at the PR. 
For reference: I have created a new NetNative targeting pack which includes the types/APIs that are the target of this change already in S.P.CoreLib. The changes on CoreFX (in this PR) are about consuming the new targeting pack, creating the two libs with forwarders to S.P.CoreLib, and packaging changes to produce a new fake TFM (uap10.1) which will help NuGet resolve the correct assets. 

@AtsushiKan you might be interested in this PR as well.

### CODE REVIEW

File: dir.props [hunk 1]
```diff
@@ -84,6 +84,28 @@
...
   </PropertyGroup>
+  <!-- temporary change until we merge dev/api to all the branches. packaging.targets defines OutOfBoxFramework items
+       which is going to be used by EnsureOOBFramework target. We need uap10.1 to be enforced as a separate target 
+       in NuGet packages. And to do that we are going to leverage EnsureOOBFramework. However, this target will fail
+       if it can't cind any compatible targets within the project.json. There are some libraries that doesn't support
+       uap10.1 compatible frameworks so defining uap10.1 within packaging.targets will break the build unless those
+       projects are annotated as not supporting uap10.1. Since uap10.1 changes are not on all corefx branches, this change
+       will cause build breaks. So Instead of defining it in packaging.targets (in buildtools), we're defining it on 
+       the branch that can handle uap10.1. Once uap10.1 is merged into the other branches we should move this OutOfBoxFramework
+       item definition into packaging.targets. -->
```
Comment (weshaggard) [inline dir.props]: Can we put this directly into the pkgproj's that we need it in instead of doing it for all packages.

File: dir.props [hunk 2]
```diff
@@ -524,6 +552,21 @@
         <NuGetTargetMoniker>WindowsPhoneApp,Version=v8.1</NuGetTargetMoniker>
       </PropertyGroup>
     </When>
+    <When Condition="'$(TargetGroup)'=='uap101aot'">
+      <PropertyGroup>
+        <PackageTargetFramework>uap10.1</PackageTargetFramework>
+        <PackageTargetRuntime>aot</PackageTargetRuntime>
+        <TargetingPackNugetPackageId>Microsoft.TargetingPack.Private.NetNative</TargetingPackNugetPackageId>
```
Comment (weshaggard) [inline dir.props]: go ahead and delete TargetingPackNugetPackageId as it is just a dead property.

**Comment by weshaggard** [comment]
>15:35:02 /mnt/resource/j/workspace/dotnet_corefx/dev_api/centos7.1_debug_prtest/build.proj(110,5): error : Dependency validation error: for Microsoft.TargetingPack.Private.CoreCLR 1.0.2-beta-24410-02 in /mnt/resource/j/workspace/dotnet_corefx/dev_api/centos7.1_debug_prtest/src/System.IO/src/project.json package prerelease is 'beta-24410-02', but expected 'beta-24412-03' for packages matching '^(?i)(Microsoft\.NETCore\.Runtime.*)|(Microsoft\.TargetingPack\.Private\.CoreCLR)$'

Looks like you have some outdated versions for CoreCLR package.

File: src/System.IO/src/System.IO.builds [hunk 3]
```diff
@@ -25,6 +25,12 @@
     <Project Include="redist\System.IO.depproj">
       <TargetGroup>netcore50aot</TargetGroup>
     </Project>
+    <Project Include="System.IO.csproj">
+      <TargetGroup>uap101aot</TargetGroup>
+    </Project>
+    <Project Include="System.IO.csproj">
```
Comment (weshaggard) [inline src/System.IO/src/System.IO.builds]: Is it necessary to introduce a uap101 non-aot build? Can we not use the other build for this?

File: src/System.IO/src/System/IO/BufferedStream.cs [hunk 4]
```diff
@@ -436,7 +436,7 @@ private int ReadFromBuffer(Byte[] array, int offset, int count, out Exception er
             }
         }
 
-        public override int Read([In, Out] byte[] array, int offset, int count)
+        public override int Read(/*[TODO: Enable after System.Private.Interop is built against S.P.CoreLib and not S.Runtime] [In, Out]*/ byte[] array, int offset, int count)
```
Comment (weshaggard) [inline src/System.IO/src/System/IO/BufferedStream.cs]: Lets make sure we track this issue.

File: dir.props (hunk 1 above)
Comment (SedarG) [inline dir.props]: Sure. It doesn't matter much which way we do it. I can change it the way you propose. 

I don't know what your plans are. But I believe there's a major packaging debt in order to have the proper resolution of UAP10.1 from the generations (NetStandard). Is there an issue tracking that?

File: src/System.IO/src/System.IO.builds (hunk 3 above)
Comment (SedarG) [inline src/System.IO/src/System.IO.builds]: I'm no sure where the binary build is going to be used (what's going to consume src/System.IO.builds). If you're saying that it's not necessary to invoke the csproj with TargetGroup=uap101, I can certainly remove it from System.IO.builds.

Updating the System.IO.Builds will not have an affect on runtime.any.system.io package. Could you please confirm that runtime.any.system.io is as you'd expect

File: dir.props (hunk 2 above)
Comment (SedarG) [inline dir.props]: will do. thanks!

File: src/System.IO/pkg/any/System.IO.pkgproj [hunk 5]
```diff
@@ -14,12 +14,19 @@
...
+    <ProjectReference Include="..\..\src\System.IO.csproj">
+      <TargetGroup>uap101</TargetGroup>
+    </ProjectReference>
 
     <!-- AOT implementation comes from AOT package -->
     <File Include="$(PlaceholderFile)">
       <TargetPath>runtimes/aot/lib/netcore50</TargetPath>
       <TargetFramework>aot</TargetFramework>
     </File>
+    <File Include="$(PlaceholderFile)">
```
Comment (ericstj) [inline src/System.IO/pkg/any/System.IO.pkgproj]: This shouldn't be needed.

**Comment by weshaggard** [comment]
LGTM

please squash before merge.

File: Packaging.props [hunk 6]
```diff
@@ -43,4 +43,4 @@
         <SkipPackageFileCheck>true</SkipPackageFileCheck>
     </File>
   </ItemGroup>
-</Project>
\ No newline at end of file
+</Project>
```
Comment (ericstj) [inline Packaging.props]: You can undo this file

**Comment by ericstj** [comment]
LGTM

**Pull request merged**